File: FlagMiner/OptionsForm.vb
Imports System.ComponentModel
Imports System.IO

Public Class OptionsForm

    Private Sub OptionsForm_FormClosing(sender As Object, e As FormClosingEventArgs) Handles Me.FormClosing
        If (DialogResult = Windows.Forms.DialogResult.None) Then e.Cancel = True
    End Sub

    Private Sub OptionsForm_Load(sender As Object, e As EventArgs) Handles Me.Load
        Me.enableCheck.Checked = Form1.options.enableCheck
        Me.enablePurge.Checked = Form1.options.enablePurge
        Me.localRepoFolder.Text = Form1.options.localRepoFolder
        Me.localSaveFolder.Text = Form1.options.localSaveFolder
        Me.markTroll.Checked = Form1.options.markTroll
        Me.RadioButton1.Checked = Form1.options.useLocal
        Me.userAgent.Text = Form1.options.userAgent
        Me.saveAndLoadFolder.Text = Form1.options.saveAndLoadFolder
        Me.deleteChildFree.Checked = Form1.options.deleteChildFree
    End Sub

    Private Sub Button2_Click(sender As Object, e As EventArgs) Handles Button2.Click
        Me.DialogResult = Windows.Forms.DialogResult.Cancel
    End Sub

    Private Sub Button1_Click(sender As Object, e As EventArgs) Handles Button1.Click
        If Not Me.ValidateChildren() Then
            Me.DialogResult = DialogResult.None
        Else
            Me.DialogResult = Windows.Forms.DialogResult.OK
        End If
    End Sub

    Private Sub localSaveFolder_Validated(sender As Object, e As EventArgs) Handles localSaveFolder.Validated
        Form1.options.localSaveFolder = Me.localSaveFolder.Text
        Form1.options.localRepoFolder = Me.localRepoFolder.Text
        Form1.options.enableCheck = Me.enableCheck.Checked
        Form1.options.enablePurge = Me.enablePurge.Checked
        Form1.options.useLocal = Me.RadioButton1.Checked
        Form1.options.markTroll = Me.markTroll.Checked
        Form1.options.userAgent = Me.userAgent.Text
        Form1.options.saveAndLoadFolder = Me.saveAndLoadFolder.Text
        Form1.options.deleteChildFree = Me.deleteChildFree.Checked
    End Sub

    Private Sub TextBox1_Validating(sender As Object, e As CancelEventArgs) Handles localSaveFolder.Validating
        If enableCheck.Checked Then
            If (localSaveFolder.Text.Length = 0) AndAlso Not Directory.Exists(localSaveFolder.Text) Then
                MsgBox("Invalid folder", MsgBoxStyle.Critical, "Flag Miner")
                e.Cancel = True
                Return
            End If
        End If
    End Sub

    Private Sub TextBox2_Validating(sender As Object, e As CancelEventArgs) Handles localRepoFolder.Validating
        If enablePurge.Checked AndAlso RadioButton1.Checked Then
            If (localRepoFolder.Text.Length = 0) AndAlso Not Directory.Exists(localRepoFolder.Text) Then
                MsgBox("Invalid folder", MsgBoxStyle.Critical, "Flag Miner")
                e.Cancel = True
                Return
            End If
        End If
    End Sub

    Private Sub userAgent_Validating(sender As Object, e As CancelEventArgs) Handles userAgent.Validating
        If (String.IsNullOrEmpty(userAgent.Text) OrElse String.IsNullOrWhiteSpace(userAgent.Text)) Then
            MsgBox("Insert a valid User Agent or press Default", MsgBoxStyle.Critical, "Flag Miner")
            e.Cancel = True
            Return
        End If
    End Sub

    Private Sub Button3_Click(sender As Object, e As EventArgs) Handles Button3.Click
        If FolderBrowserDialog1.ShowDialog = Windows.Forms.DialogResult.OK Then
            localSaveFolder.Text = FolderBrowserDialog1.SelectedPath
        End If
    End Sub

    Private Sub CheckBox1_CheckedChanged(sender As Object, e As EventArgs) Handles enablePurge.CheckedChanged
        Panel1.Enabled = enablePurge.Checked
    End Sub

    Private Sub Button4_Click(sender As Object, e As EventArgs) Handles Button4.Click
        If FolderBrowserDialog1.ShowDialog = Windows.Forms.DialogResult.OK Then
            localRepoFolder.Text = FolderBrowserDialog1.SelectedPath
        End If
    End Sub

    Private Sub Button5_Click(sender As Object, e As EventArgs) Handles Button5.Click
        userAgent.Text = Form1.DefaultUserAgent
    End Sub

    Private Sub Button6_Click(sender As Object, e As EventArgs) Handles Button6.Click
        If FolderBrowserDialog1.ShowDialog = Windows.Forms.DialogResult.OK Then
            saveAndLoadFolder.Text = FolderBrowserDialog1.SelectedPath
        End If
    End Sub

End Class

File: FlagMiner/OptionsForm.cs
using System;
using System.Collections.Generic;
using System.Windows.Forms;
using System.Linq;
using System.ComponentModel;
using System.IO;
namespace FlagMiner
{

    partial class OptionsForm : Form
	{
        private FlagMiner myForm1;
        public OptionsForm(FlagMiner frm1)
        {
            myForm1 = frm1;
            Load += OptionsForm_Load;
            FormClosing += OptionsForm_FormClosing;
            InitializeComponent();
        }
		private void OptionsForm_FormClosing(object sender, FormClosingEventArgs e)
		{
			if (DialogResult == System.Windows.Forms.DialogResult.None)
				e.Cancel = true;
		}

		private void OptionsForm_Load(object sender, EventArgs e)
		{
            var options = OptionsManager.OptionsInstance;

            this.BackendServersTextBox.Clear();
            if (options.backendServers != null)
            {
                foreach (string st in options.backendServers)
                {
                    if (st != "")
                        this.BackendServersTextBox.Text += st + "\n";
                }
            }

            this.LocalFlagSaveFolderTextBox.Text = options.localSaveFolder;
            this.LocalRepoFolderTextBox.Text = options.localRepoFolder;

            this.EnableCheckCheckBox.Checked = options.enableCheck;
            this.enableCheck_CheckedChanged(null, System.EventArgs.Empty);
            this.EnablePurgeCheckBox.Checked = options.enablePurge;
            this.enablePurge_CheckedChanged(null, System.EventArgs.Empty);
            this.UseLocalRepoRadioButton.Checked = options.useLocal;
            this.MarkTrollCheckBox.Checked = options.markTroll;
            this.DeleteChildFreeCheckBox.Checked = options.deleteChildFree;

            this.UserAgentTextBox.Text = options.userAgent;

            this.TreeSaveAndLoadFolderTextBox.Text = options.saveAndLoadFolder;

            this.RepoUrlTextBox.Text = options.repoUrl;
		}

		private void Button2_Click(object sender, EventArgs e)
		{
			this.DialogResult = System.Windows.Forms.DialogResult.Cancel;
		}

		private void Button1_Click(object sender, EventArgs e)
		{
			if (!this.ValidateChildren()) {
				this.DialogResult = DialogResult.None;
			} else {
                OptionsForm_UpdateOpts();
				this.DialogResult = System.Windows.Forms.DialogResult.OK;
			}
		}

		private void OptionsForm_UpdateOpts()
		{
            var options = OptionsManager.OptionsInstance;

            options.backendServers = new List<string>();
            foreach (string st in this.BackendServersTextBox.Lines)
            {
                if (st != "")
                { options.backendServers.Add(st); }
            }

            options.localSaveFolder = this.LocalFlagSaveFolderTextBox.Text;
            options.localRepoFolder = this.LocalRepoFolderTextBox.Text;

            options.enableCheck = this.EnableCheckCheckBox.Checked;
            options.enablePurge = this.EnablePurgeCheckBox.Checked;
            options.useLocal = this.UseLocalRepoRadioButton.Checked;
            options.markTroll = this.MarkTrollCheckBox.Checked;
            options.deleteChildFree = this.DeleteChildFreeCheckBox.Checked;

            options.userAgent = this.UserAgentTextBox.Text;

            options.saveAndLoadFolder = this.TreeSaveAndLoadFolderTextBox.Text;

            options.repoUrl = this.RepoUrlTextBox.Text;

            OptionsManager.OptionsInstance = options;
            OptionsManager.SaveOptions();
		}

		private void TextBox1_Validating(object sender, CancelEventArgs e)
		{
			if (EnableCheckCheckBox.Checked && (LocalFlagSaveFolderTextBox.Text.Length == 0 || !Directory.Exists(LocalFlagSaveFolderTextBox.Text))) {
                MessageBox.Show("Invalid local save folder.", "Flag Miner", MessageBoxButtons.OK, MessageBoxIcon.Error);
				e.Cancel = true;
				return;
			}
		}

		private void TextBox2_Validating(object sender, CancelEventArgs e)
		{
			if (EnablePurgeCheckBox.Checked && UseLocalRepoRadioButton.Checked && (LocalRepoFolderTextBox.Text.Length == 0) || !Directory.Exists(LocalRepoFolderTextBox.Text)) {
                MessageBox.Show("Invalid local repo folder.", "Flag Miner", MessageBoxButtons.OK, MessageBoxIcon.Error);
				e.Cancel = true;
				return;
			}
		}

		private void userAgent_Validating(object sender, CancelEventArgs e)
		{
			if ((string.IsNullOrEmpty(UserAgentTextBox.Text) || string.IsNullOrWhiteSpace(UserAgentTextBox.Text))) {
                MessageBox.Show("Insert a valid User Agent or press Default", "Flag Miner", MessageBoxButtons.OK, MessageBoxIcon.Error);
				e.Cancel = true;
				return;
			}
		}

        private void repoUrl_Validating(object sender, CancelEventArgs e)
        {
            if ((string.IsNullOrEmpty(RepoUrlTextBox.Text) || string.IsNullOrWhiteSpace(RepoUrlTextBox.Text)))
            {
                MessageBox.Show("Insert a valid Url for the Repository or press Default", "Flag Miner", MessageBoxButtons.OK, MessageBoxIcon.Error);
                e.Cancel = true;
                return;
            }
        }

        private void backendServers_Validating(object sender, CancelEventArgs e)
        {
            if (this.BackendServersTextBox.Lines.All(s => string.IsNullOrWhiteSpace(s)))
            {
                MessageBox.Show("Insert a valid Url for the backend server or press Default", "Flag Miner", MessageBoxButtons.OK, MessageBoxIcon.Error);
                e.Cancel = true;
                return;
            }
        }

        private void Button3_Click(object sender, EventArgs e)
		{
			if (FolderBrowserDialog1.ShowDialog() == System.Windows.Forms.DialogResult.OK) {
				LocalFlagSaveFolderTextBox.Text = FolderBrowserDialog1.SelectedPath;
			}
		}

		private void enablePurge_CheckedChanged(object sender, EventArgs e)
		{
			Panel1.Enabled = EnablePurgeCheckBox.Checked;
		}

		private void Button4_Click(object sender, EventArgs e)
		{
			if (FolderBrowserDialog1.ShowDialog() == System.Windows.Forms.DialogResult.OK) {
				LocalRepoFolderTextBox.Text = FolderBrowserDialog1.SelectedPath;
			}
		}

		private void defaultUserAgentButton_Click(object sender, EventArgs e)
		{
			UserAgentTextBox.Text = Properties.Resources.DefaultUserAgent;
        }

		private void Button6_Click(object sender, EventArgs e)
		{
			if (FolderBrowserDialog1.ShowDialog() == System.Windows.Forms.DialogResult.OK) {
				TreeSaveAndLoadFolderTextBox.Text = FolderBrowserDialog1.SelectedPath;
			}
		}

        private void enableCheck_CheckedChanged(object sender, EventArgs e)
        {
            selectLocalDestFolderButton.Enabled = EnableCheckCheckBox.Checked;
        }

        private void defaultBackendButton_Click(object sender, EventArgs e)
        {
            if (MessageBox.Show("Restore default server address?", "Flag Miner", MessageBoxButtons.YesNo) == System.Windows.Forms.DialogResult.Yes)
            {
                this.BackendServersTextBox.Clear();
                // DEFAULT SERVERS
                this.BackendServersTextBox.Text += Properties.Resources.backendBaseUrl + "\n";
            }
        }

        private void defaultRepoUrlButton_Click(object sender, EventArgs e)
        {
            this.RepoUrlTextBox.Text = Properties.Resources.DefaultflegsBaseUrl;
        }

        public override bool ValidateChildren()
        {
            foreach (Control control in new List<Control> { this.BackendServersTextBox, this.LocalRepoFolderTextBox, this.RepoUrlTextBox, this.UserAgentTextBox, this.LocalFlagSaveFolderTextBox })
            {
                control.Focus();
                if (!Validate())
                {
                    return false;
                }
            }
            return true;
        }

	}
}
